### DISCUSSION TITLE
Why `toBeVisible` doesn't work with `getByRole`?

### DISCUSSION THREAD

**Discussion by felipecesr** [testing-library/jest-dom#622] [General]
Hi guys! Recently I needed to use the `toBeVisible` assertion with an element that I got using `getByRole`.
```
expect(getByRole('button', { name:  'Visible Example' })).toBeVisible()
```
It doesn't work. I read the docs and all the examples are using `ByText` queries.
```
expect(getByText('Visible Example')).toBeVisible()
```

Are there any reasons for that?

**Comment by gnapse** [comment]
The fact that it works or not is most likely unrelated to whether you used `getByRole` or `getByText` to find the element. `toBeVisible` has no knowledge of what means were used to query for that element. It merely receives the element found by whatever means, and asserts that it is visible or not.

The documentation only using `getByText` is also irrelevant. The documentation in our README uses placeholder names for helpers, but in practice those helpers are part of a different library (react-testing-library). And also, it is likely that when our README was written, `getByRole` did not exist yet. It was not part of react-testing-library at first.

Lastly, if you can find a situation where using a different query function changes the outcome of `toBeVisible`, then that sounds more like a bug. Is that what you've found?

**Reply by felipecesr** [reply]
Yes! When I use `getByText` I can find the element, but when I use `getByRole` I can't find the element.

**Reply by gnapse** [reply]
Then, it might not have anything to do with `toBeVisible`. I bet it is failing because `getByText` finds the element, whereas `getByRole` does not find it. But that has nothing to do with `toBeVisible`.

One reason why `getByText` can find an element while `getByRole` does not, could be because the element might be hidden from the accessibility tree (e.g. using `aria-hidden`).

Can you please provide a reproducible environment? One that clearly replicates the issue, say, in a codesandbox repository?

**Reply by felipecesr** [reply]
Sure! You can fork this [codesandbox repository](https://codesandbox.io/p/devbox/9klqhg?migrateFrom=f58s8c&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clzrevjxz00063b835kb5954x%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clzrevjxz00023b83wemb1m0v%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clzrevjxz00043b83hbz692vh%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clzrevjxz00053b837wkc44kg%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clzrevjxz00023b83wemb1m0v%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzrevjxz00013b83fi6aaps9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%257D%255D%252C%2522id%2522%253A%2522clzrevjxz00023b83wemb1m0v%2522%252C%2522activeTabId%2522%253A%2522clzrevjxz00013b83fi6aaps9%2522%257D%252C%2522clzrevjxz00053b837wkc44kg%2522%253A%257B%2522id%2522%253A%2522clzrevjxz00053b837wkc44kg%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clzrevjxz00043b83hbz692vh%2522%253A%257B%2522id%2522%253A%2522clzrevjxz00043b83hbz692vh%2522%252C%2522activeTabId%2522%253A%2522clzrfj1pu00ru3b83gfeopt6e%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzrfj1pu00ru3b83gfeopt6e%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clzrfj1u8002sd9f87kpq0oi9%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) and run `pnpm test`. Let me know if you have any questions.

**Reply by gnapse** [reply]
Exactly. This confirms what I say. `toBeVisible` has nothing to do with this. It's the `getBy*` function used what changes the behavior.

- `getByText` searches the DOM without any other consideration about whether the element should be included in the search or not. Since the element exists, it returns it.
- `getByRole` takes into account a lot more stuff, including being "blind" to elements that do not exist from the accessibility tree POV. Being an element with the `display: none` style, `getByRole` does not consider it, and fails the search, returning `null`.

So this is not a "`toBeVisible` does not work with `getByRole`". This is more like `.toBeVisible` returns false if it is given `null` for input instead of an element.

**Reply by felipecesr** [reply]
So this is the expected behavior for the `ByRole` query, right? It shouldn't consider invisible elements like that, even with `hidden: true`, right?

**Reply by gnapse** [reply]
I think so, but I'm not a 100% sure. Also, those helpers are not part of jest-dom. I suggest you ask in https://github.com/testing-library/react-testing-library or https://github.com/testing-library/dom-testing-library, which is where these DOM querying helpers come from.

**Reply by felipecesr** [reply]
Ok! Thank you for your help.
